File: src/data/courses/dsa/dynamic-programming/string-dp/longest-common-subsequence/code/python/longest_common_subsequence.py
class Solution:
    def longestCommonSubsequence(self, t1, t2) -> int:
        m,n=len(t1),len(t2); dp=[0]*(n+1)
        for i in range(1,m+1):
            prev=0
            for j in range(1,n+1):
                temp=dp[j]
                dp[j]=prev+1 if t1[i-1]==t2[j-1] else max(dp[j],dp[j-1])
                prev=temp
        return dp[n]

File: src/data/courses/dsa/dynamic-programming/string-dp/longest-common-subsequence/code/cpp/longest_common_subsequence.cpp
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
class Solution {
public:
    int longestCommonSubsequence(string t1, string t2) {
        int m=t1.size(),n=t2.size();
        vector<int> dp(n+1,0);
        for(int i=1;i<=m;i++){
            int prev=0;
            for(int j=1;j<=n;j++){
                int temp=dp[j];
                dp[j]=t1[i-1]==t2[j-1]?prev+1:max(dp[j],dp[j-1]);
                prev=temp;
            }
        }
        return dp[n];
    }
};